client batch requests

niklasad1 opened the pull request:
Closes #103

Needs to updated to work on WebSockets too before it's ready to go.

niklasad1 commented on http-client/src/client.rs:
```diff
@@ -23,6 +22,71 @@ impl HttpClient {
+	/// Returns `Error` if any of the requests fails.
+	//
+	// TODO(niklasad1): maybe simplify generic `requests`, it's quite unreadable.
+	pub async fn batch_request<'a, T>(
```
move this to client trait?!

maciejhirsz commented on ws-client/src/manager.rs:
```diff
@@ -53,6 +54,9 @@ pub struct RequestManager {
 	subscriptions: HashMap<SubscriptionId, RequestId>,
+	/// Pending batch requests
+	// NOTE: BTreeSet to get sorted order of requestIDs.
+	batches: HashMap<BTreeSet<RequestId>, (Vec<RequestId>, PendingBatchOneshot)>,
```
`FnvHashMap<Vec<RequestId>, (Vec<RequestId>, PendingBatchOneshot)>,` and sort the Ids before accessing the map.

dvdplm commented:
> as the `server` is free to process the batch in any order such that the client has try any order of the request IDs.

The spec says "The Response objects being returned from a batch call MAY be returned in any order within the Array. The Client SHOULD match contexts between the set of Request objects and the resulting set of Response objects based on the id member within each Object." – but here you implement an ordered set of responses. Can you elaborate on this? What do they mean by "SHOULD match contexts"?

dvdplm commented on ws-client/src/client.rs:
```diff
@@ -375,8 +420,50 @@
+						None => unreachable!("All request ID's valid checked by RequestManager above; qed"),
+					};
+					// Dummy value should never be.
+					let (_, rp) = std::mem::replace(&mut rps_unordered[pos], (0, JsonValue::Number(0.into())));
```
Is there a guarantee that `Id` is non-zero? I took a glance and it seems that it boils down to a `u64` so using 0 as the tombstone is perhaps not 100% safe?

niklasad1 replied:
This is just dummy value because each request IDs should be unique and read exactly once, ensured by the RequestManager if we don't have any bugs :)
 
So think of this as `std::mem::take` but I could instead clone the value if you like it better.

dvdplm replied:
`std::mem::replace` is fine. My concern was more that we could end up with a valid ID of `0` and that might be bad? We are using `0` to signal "free" and the comment above, "Dummy value should never be", made me think `0` must never be used for an ID. If this is the case then we should use the type system to ensure IDs are non-zero.

niklasad1 replied:
Switched to `Vec::swap_remove` it is `O(1)` but doesn't preserve order.

dvdplm reviewed (commented):
It seems like 90% of the complexity here comes from the requirement to return responses in the order they were submitted. As noted above I'm not 100% convinced this is what the spec actually says so before commenting further I'd like to be convinced I'm wrong.

niklasad1 commented:
> as the `server` is free to process the batch in any order such that the client has try any order of the request IDs.

I meant that each "request/request object" in the batch may be returned by server in any order, clarified the PR description.
I think the tests I implemented show this quite clearly.

> The spec says "The Response objects being returned from a batch call MAY be returned in any order within the Array. The Client SHOULD match contexts between the set of Request objects and the resulting set of Response objects based on the id member within each Object." – but here you implement an ordered set of responses. Can you elaborate on this? What do they mean by "SHOULD match contexts"?

I interpreted it as each request ID should be matched with corresponding request ID in the response, or in their words "match contexts".

Technically, I implemented it in the same order as the requests are performed by the user.
When we receive a batch (internally a vec of requests), then first we copy to vec into a BTreeSet (which is ordered) and store orignal order in another Vec:

```rust
batches: HashMap<BTreeSet<RequestId>, (Vec<RequestId>, PendingBatchOneshot)>
```

This is just to avoid trying out all possible combinations of the request ID and the same could be achieved by sorting the `Vec<RequestId>` and put them in the HashMap as Maciej pointed out.

As I told Maciej I will try to provide some benchmarks on this

dvdplm commented:
@niklasad1 brainfart on my part; I was thinking that "client" here meant the user code that is sending the batch (which is completely wrong). Ignore me. :)

niklasad1 commented on http-client/src/client.rs:
```diff
@@ -23,6 +22,71 @@ impl HttpClient {
+	/// Returns `Ok` if all requests were answered successfully.
+	/// Returns `Error` if any of the requests fails.
+	//
+	// TODO(niklasad1): maybe simplify generic `requests`, it's quite unreadable.
```
`requests: impl IntoIterator<Item = (impl Into<String>, impl Into<jsonrpc::Params>)>,`

Was what I referred to in the TODO must people will probably use a `Vec` and `IntoIterator` allows unordered collections such as HashMap which might be a footgun in the case.

So maybe accepting requests as a `slice` is better.

dvdplm replied:
I like the idea of using a slice. @maciejhirsz thoughts?

niklasad1 commented on http-client/src/client.rs:
```diff
@@ -23,6 +22,71 @@ impl HttpClient {
+			.map_err(|e| Error::TransportError(Box::new(e)))?;
+
+		match response {
+			jsonrpc::Response::Single(_) => {
```
This means that response was `deserialized` as single request object i.e, not batch (which has to contain an array of requests with a at least one request) which doesn't follow the spec.

So, we return a error for this instead of panicking.

niklasad1 commented on http-client/src/client.rs:
```diff
@@ -23,6 +22,71 @@ impl HttpClient {
+					json_responses[pos] = json_val;
+				}
+				let responses: Result<_, _> =
+					json_responses.into_iter().map(|val| jsonrpc::from_value(val).map_err(Error::ParseError)).collect();
```
yeah, I think it did like this because `jsonrpc::from_value` returns `T.

Then we either to add `T: Default` or put in an `Option` where the json_responses is declared.

dvdplm replied:
`T: Default` sounds reasonable to me.

dvdplm commented:
> Further, batch requests does consume a `request ID` for each request in the batch and it "might violate" the "max_concurrent_request_limit" depending on how we define that I suppose. For example if a user sets this to 10 and then perform a batch request of 10, then no more requests are allowed until the batch has been completed. So question is that if a batch request should be regarded as one request or the number of requests in the batch

I think a batch request containing `n` requests should count against the `max_concurrent_request_limit` as `n`. In my head the `max_concurrent_request_limit` setting should default to a fairly high value – 1000? hundreds? – and it should be thought of as a safety valve and not really as a throttling mechanism.

niklasad1 commented on http-client/src/client.rs:
```diff
@@ -79,4 +78,62 @@ impl Client for HttpClient {
+	{
+		let mut calls = Vec::with_capacity(batch.len());
+		// NOTE(niklasad1): `ID` is not necessarily monotonically increasing.
+		let mut ordered_requests = Vec::with_capacity(batch.len());
```
I realized that we could have an additional request_set `HashSet<RequestID>`  here to do quick lookups to avoid O(n) lookups

dvdplm commented on http-client/src/client.rs:
```diff
@@ -88,8 +89,9 @@ impl Client for HttpClient {
 		let mut calls = Vec::with_capacity(batch.len());
 		// NOTE(niklasad1): `ID` is not necessarily monotonically increasing.
 		let mut ordered_requests = Vec::with_capacity(batch.len());
+		let mut request_set = FnvHashMap::with_capacity_and_hasher(batch.len(), Default::default());
```
I wonder why they didn't keep `with_capacity()`.

niklasad1 replied:
:shrug:

niklasad1 commented:
> I think a batch request containing n requests should count against the max_concurrent_request_limit as n. In my head the max_concurrent_request_limit setting should default to a fairly high value – 1000? hundreds? – and it should be thought of as a safety valve and not really as a throttling mechanism.

Indeed and this makes it more complicated as request ID and batch ID (request ID of request within the batch) needs to be distinguished. So, I'm not super excited about this and it's error prone to have return the request ID after it has been completed.

Maybe better just with a simple counter and guard the hashmap/other collections when inserting new requests.

niklasad1 commented on ws-client/src/manager.rs:
```diff
@@ -38,33 +38,63 @@ pub enum RequestStatus {
+/// Batch state.
+pub struct BatchState {
+	/// Order that the request was performed in.
+	pub order: FnvHashMap<BatchId, usize>,
```
I added this to avoid the O(n) searches.

dvdplm commented on ws-client/src/client.rs:
```diff
@@ -374,8 +411,50 @@ async fn background_task(
+					rps_unordered.push((id, rp));
+				}
+
+				digest.sort();
```
Github Actions says "used sort instead of sort_unstable to sort primitive type `u64`" here. Do you see those warnings too or is it just me?

dvdplm commented on ws-client/src/jsonrpc_transport.rs:
```diff
@@ -31,6 +31,45 @@ impl Sender {
+			ids.push(batch_id);
+			calls.push(jsonrpc::Call::MethodCall(jsonrpc::MethodCall {
+				jsonrpc: jsonrpc::Version::V2,
+				method: method.into(),
```
GH Actions says "useless conversion to the same type: `std::string::String`"

dvdplm reviewed (commented):
> So, I'm not super excited about this and it's error prone to have return the request ID after each request has been completed.

FWIW it's not something that should block this PR. Let's log an issue and move on for now.

niklasad1 commented:
re benches:

I tried against `jsonrpc servers` because our `WebSocket server` doesn't support batches and funny enough I can't compile with `tokio 0.2` for tests because of our `test-util crate` uses `hyper 14` ^^

However, I did some manual wallclock benches

```
HTTP 10 requests in a batch took 0 ms
HTTP 100 requests in a batch took 0 ms
HTTP 1000 requests in a batch took 3 ms
HTTP 10000 requests in a batch took 35 ms
WebSocket 10 requests in a batch took 4 ms
WebSocket 100 requests in a batch took 0 ms
WebSocket 1000 requests in a batch took 3 ms
WebSocket 10000 requests in a batch took 34 ms
```

Looks alright, I think 48 concurrent HTTP requests took ~1ms on my machine

niklasad1 commented on types/src/traits.rs:
```diff
@@ -19,6 +19,18 @@ pub trait Client {
+	/// Returns `Error` if any of the requests in batch fails.
+	async fn batch_request<T, M, P>(&self, batch: Vec<(M, P)>) -> Result<Vec<T>, Error>
+	where
+		T: DeserializeOwned + Default + Clone,
```
`Default + Clone` is required to do `vec![T::default(); len]`

niklasad1 commented on http-client/src/client.rs:
```diff
@@ -79,4 +79,62 @@ impl Client for HttpClient {
 		jsonrpc::from_value(json_value).map_err(Error::ParseError)
 	}
+
+	async fn batch_request<T, M, P>(&self, batch: Vec<(M, P)>) -> Result<Vec<T>, Error>
```
I think it would be possible to have:

```rust
	async fn batch_request<T, M, P>(&self, batch: &[(M, P)]) -> Result<Vec<T>, Error>
	where
		T: DeserializeOwned + Default + Clone,
		M: Into<String> + Send,
		P: Into<Params> + Send + Sync,
		for<'a> &'a M: Into<String> + Send,
		for<'a> &'a P: Into<Params> + Send;
```

It becomes super complicated.

dvdplm replied:
Fair enough. Let's use a `Vec` then, keep it simple and opinionated.

niklasad1 commented on ws-client/src/jsonrpc_transport.rs:
```diff
@@ -31,6 +31,45 @@ impl Sender {
+
+		if let Err(send_back) = request_manager.insert_pending_batch(ids, batch.send_back, req_id) {
+			request_manager.reclaim_request_id(req_id);
+			let _ = send_back.send(Err(Error::InvalidRequestId));
```
maybe a new error batches?

dvdplm commented on types/src/error.rs:
```diff
@@ -30,6 +30,9 @@ pub enum Error {
 	#[error("Frontend/backend channel error: {0}")]
 	Internal(#[source] futures::channel::mpsc::SendError),
+	/// Invalid response,
+	#[error("Invalid response detected: {0}")]
```
```suggestion
	#[error("Invalid response: {0}")]
```
